src/commands/up.rs

```diff
@@ -646,6 +662,81 @@ fn trigger_command_for_resolved_app_source(
         .collect()
 }
 
+/// Scrubs the locked app to only contain the given list of components
+/// Introspects the LockedApp to find and selectively retain the triggers that correspond to those components
+fn retain_components(locked_app: &mut LockedApp, retained_components: &[String]) -> Result<()> {
+    // Create a temporary app to access parsed component and trigger information
+    let tmp_app = spin_app::App::new("tmp", locked_app.clone());
+    validate_retained_components_exist(&tmp_app, retained_components)?;
+    validate_retained_components_service_chaining(&tmp_app, retained_components)?;
+    let (component_ids, trigger_ids): (HashSet<String>, HashSet<String>) = tmp_app
+        .triggers()
+        .filter_map(|t| match t.component() {
+            Ok(comp) if retained_components.contains(&comp.id().to_string()) => {
+                Some((comp.id().to_owned(), t.id().to_owned()))
+            }
+            _ => None,
+        })
+        .collect();
+    locked_app
+        .components
+        .retain(|c| component_ids.contains(&c.id));
+    locked_app.triggers.retain(|t| trigger_ids.contains(&t.id));
+    Ok(())
+}
+
+/// Validates that all service chaining of an app will be satisfied by the retained components.
+/// All wildcard service chaining is disallowed
+fn validate_retained_components_service_chaining(
+    app: &spin_app::App,
+    retained_components: &[String],
+) -> Result<()> {
+    app
+        .triggers().try_for_each(|t| {
+        if let Ok(component) = t.component() {
+            if retained_components.contains(&component.id().to_string()) {
+            let allowed_hosts = allowed_outbound_hosts(&component).context("failed to get allowed hosts")?;
+            allowed_hosts.iter().try_for_each(|host| {
+                // Templated URLs are not resolved at this point, so ignore unresolvable URIs
+                if let Ok(uri) = host.parse::<http::Uri>() {
```

Comment:
Allowed host configs are very often not valid Uris but are still statically resolvable. For example, `*://{foo.spin.internal, bar.spin.internal}`. This check wouldn't run because the above cannot be parsed as an `http::Uri`. 

You might want to consider `AllowedHostConfig::parse` instead. This handles all of the interpolation syntax that allowed host configs are allowed to have. You can then add a new method like `AllowedHostConfig::service_chaining_target` that returns a `Vec<String>` with all of the components that that `AllowedHostConfig` targets.

Reply:
I didn't know you could chain to multiple targets in one host `*://{foo.spin.internal, bar.spin.internal}`. Should we update the docs on this: https://developer.fermyon.com/spin/v2/http-outbound#local-service-chaining?